chalk-solve/src/display/state.rs

```diff
@@ -71,35 +79,134 @@ impl<T: Copy + Ord> DefIdAliases<T> {
     }
 }
 
+#[derive(Debug)]
+struct IdAliases<I: Interner> {
+    id_aliases: IdAliasStore<UnifiedId<I>>,
+}
+
+impl<I: Interner> Default for IdAliases<I> {
+    fn default() -> Self {
+        IdAliases {
+            id_aliases: IdAliasStore::default(),
+        }
+    }
+}
+
+/// Writer state which persists across multiple writes.
+///
+/// Currently, this means keeping track of what IDs have been given what names,
+/// including deduplication information.
+///
+/// This data is stored using interior mutability - clones will point to the same underlying
+/// data.
+///
+/// Uses a separate type, `P`, for the database stored inside to account for
+/// `Arc` or wrapping other storage mediums.
+#[derive(Debug)]
+pub struct WriterState<I, DB: ?Sized, P = DB>
+where
+    DB: RustIrDatabase<I>,
+    P: Borrow<DB>,
+    I: Interner,
+{
+    pub(super) db: P,
+    id_aliases: Arc<Mutex<IdAliases<I>>>,
+    _phantom: PhantomData<DB>,
+}
+
+impl<I, DB: ?Sized, P> Clone for WriterState<I, DB, P>
+where
+    DB: RustIrDatabase<I>,
+    P: Borrow<DB> + Clone,
+    I: Interner,
+{
+    fn clone(&self) -> Self {
+        WriterState {
+            db: self.db.clone(),
+            id_aliases: self.id_aliases.clone(),
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<I, DB: ?Sized, P> WriterState<I, DB, P>
+where
+    DB: RustIrDatabase<I>,
+    P: Borrow<DB>,
+    I: Interner,
+{
+    pub fn new(db: P) -> Self {
+        WriterState {
+            db,
+            id_aliases: Arc::new(Mutex::new(IdAliases::default())),
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Returns a new version of self containing a wrapped database which
+    /// references the outer data.
+    ///
+    /// `f` will be run on the internal database, and the returned result will
+    /// wrap the result from `f`. For consistency, `f` should always contain the
+    /// given database, and must keep the same ID<->item relationships.  
+    pub(super) fn wrap_db_ref<'a, DB2: ?Sized, P2, F>(&'a self, f: F) -> WriterState<I, DB2, P2>
+    where
+        P: 'a,
```

Comment:
This bound doesn't seem to be needed

Reply:
I think we need it because `f` takes `&'a P`, that being sound requires `P: 'a`, and since we're just introducing `'a` here, there's no guarantee it really is shorter than `P`?

Specifically, we get:

```
error[E0495]: cannot infer an appropriate lifetime for borrow expression due to conflicting requirements
   --> chalk-solve\src\display\state.rs:160:19
    |
160 |             db: f(&self.db),
    |                   ^^^^^^^^
    |
note: first, the lifetime cannot outlive the anonymous lifetime #1 defined on the method body at 152:5...
   --> chalk-solve\src\display\state.rs:152:5
    |
152 | /     pub(super) fn wrap_db_ref<'a, DB2: ?Sized, P2, F>(&self, f: F) -> WriterState<I, DB2, P2>
153 | |     where
154 | |         P: 'a,
155 | |         DB2: RustIrDatabase<I>,
...   |
163 | |         }
164 | |     }
    | |_____^
note: ...so that reference does not outlive borrowed content
   --> chalk-solve\src\display\state.rs:160:19
    |
160 |             db: f(&self.db),
    |                   ^^^^^^^^
note: but, the lifetime must be valid for the lifetime `'a` as defined on the method body at 152:31...
   --> chalk-solve\src\display\state.rs:152:31
    |
152 |     pub(super) fn wrap_db_ref<'a, DB2: ?Sized, P2, F>(&self, f: F) -> WriterState<I, DB2, P2>
    |                               ^^
    |
160 |             db: f(&self.db),
    |                   ^^^^^^^^

error: aborting due to previous error
```

We need `'a` at all, because we want to be able to create a `P2` which directly wraps `P`, and in order to do that the `&P` needs a lifetime that's guaranteed to outlive the function. If it was just `FnOnce(&P) -> P2`, then the `&P` could have been created inside `wrap_db_ref`, and then it really would be unsound to have `P2` wrapping `P`.